src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/ProjectNames.cs

```diff
@@ -107,6 +110,66 @@ private static string GetProjectGuid(string fullname, IVsSolution5 vsSolution5)
             return guid.ToString();
         }
 
+        /// <summary>
+        /// Factory method initializing instance of <see cref="ProjectNames"/> with values retrieved from <see cref="IVsSolution2"/> and <see cref="IVsHierarchy"/>.
+        /// </summary>
+        /// <param name="fullPath">Full path to the project file</param>
+        /// <param name="vsSolution2">Instance of <see cref="IVsSolution2"/></param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation</param>
+        /// <returns></returns>
+        public static async Task<ProjectNames> FromIVsSolution2(string fullPath, IVsSolution2 vsSolution2, CancellationToken cancellationToken)
+        {
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            ErrorHandler.ThrowOnFailure(vsSolution2.GetProjectOfUniqueName(fullPath, out IVsHierarchy project));
+            ErrorHandler.ThrowOnFailure(vsSolution2.GetGuidOfProject(project, out Guid guid));
+            ErrorHandler.ThrowOnFailure(vsSolution2.GetUniqueNameOfProject(project, out string uniqueName));
+            ErrorHandler.ThrowOnFailure(project.GetProperty((uint)VSConstants.VSITEMID.Root, (int)__VSHPROPID.VSHPROPID_Name, out object projectNameObject));
+            string shortName = (string)projectNameObject;
```

Comment:
Additionally, if the project system/IVsSolution is reporting a successful invocation, but the value is null or empty, it suggests that is the actual value for the project, and therefore isn't an error condition. In which case `ProjectNames` will "incorrectly" throw, but unless we have telemetry showing that we're failing because of this very specific scenario, I feel it's overengineering to try to go too much out of our way to check for error conditions that we currently assume are not possible.